### DISCUSSION TITLE
Experiencing high cpu and freezes when triggering code actions

### DISCUSSION THREAD

**Discussion by dotfrag** [LazyVim/LazyVim#3948] [Q&A] [answered]
I'm experiencing high CPU usage and freezes (without any recovery) if I trigger Code Actions (probably any vim.ui.select?) before `fzf-lua` has loaded.

![2024-07-07T12:13:11,150954945+03:00](https://github.com/LazyVim/LazyVim/assets/17456867/924d63b6-b456-4dbc-bb84-75ef0754edc4)

I believe it started happening after https://github.com/LazyVim/LazyVim/commit/9acadc7ec289966cf8f4f6c37c348f286ed73ffe.

This is a problem with my config because I can't replicate it using `repro.lua`. Specifically, the [fzf spec](https://github.com/dotfrag/nvim/blob/main/lua/plugins/fzf.lua). If I remove it, it doesn't happen. The issue is somewhere in the `keys`, but I don't understand what. Any help is appreciated.

Update: if I checkout LazyVim to `v12.23.0`, before https://github.com/LazyVim/LazyVim/commit/b9a8a4ad7f39c8ed3f4febb25dabc9ae24110f98 and https://github.com/LazyVim/LazyVim/commit/9acadc7ec289966cf8f4f6c37c348f286ed73ffe, this doesn't happen.

Update 2: It seems related to the require statements for `fd_opts` and `rg_opts`, probably some kind of race condition between `init` and the `require` statement in `keys`, causing an infinite loop?

**Comment by dpetka2001** [comment, marked as answer]
Could you try this instead and see how it behaves?
```lua
    local config = require("fzf-lua.config")
    local fd_opts = config.defaults.files.fd_opts
    local rg_opts = config.defaults.grep.rg_opts
```

**Reply by dotfrag** [reply]
Hey, that worked! How did you come up with it? Is it because the `defaults` are loaded later?

**Reply by dpetka2001** [reply]
I just went along with how LazyVim did it. It uses `require("fzf-lua.config")` to get access to the default values. It was just a lucky guess. I still can't explain the difference between the 2 because I'm not that knowledgeable. 

My guess would be that `require("fzf-lua").defaults...` tries to load the module and access the `defaults` immediately. Whereas, `require("fzf-lua.config").defaults....` accesses the defaults through the `config` module (it's copying the `fzf-lua.defaults` into `fzf-lua.config.defaults` as can be seen [here](https://github.com/ibhagwan/fzf-lua/blob/30e8e962405c274d7766dda207e1761a0105929c/lua/fzf-lua/config.lua#L128-L129)) which I'm guessing is available after the `fzf-lua` loads. It's like some kind of `vim.schedule`/`vim.defer_fn` kind of thing (again I'm just guessing here). 

I would also be interested to know what is really happening, so if you don't mind pinging Folke to give us an explanation whenever he has some free time, I would welcome it :smile:

**Reply by dpetka2001** [reply]
I believe if you did `require("fzf-lua.defaults").defaults.files.fd_opts` it might have also worked if you don't mind trying.

**Reply by dpetka2001** [reply]
When you did `require("fzf-lua").defaults` you essentially load `fzf-lua` and access the `defaults`. That may be causing some conflict with `lazy.nvim`'s setup function with `fzf-lua`'s lazy-loading. But when you do either `require("fzf-lua.config").defaults...` or `require("fzf-lua.defaults").defaults...` you define that you access the specific module after `fzf-lua` has been loaded.

**Reply by dpetka2001** [reply]
My final guess would be that `require("fzf-lua").defaults` accesses the `M.defaults = require("fzf-lua.defaults").defaults` which should be available after the plugin has loaded. The second approach `require("fzf-lua.config").defaults` or `require("fzf-lua.defaults").defaults` gives you direct access to the table of the specific module, which is just the original default values before any modification.

Somewhere along these thoughts of mine (including my previous comments) must lie the truth :stuck_out_tongue:

**Reply by dotfrag** [reply]
> My guess would be that require("fzf-lua").defaults... tries to load the module and access the defaults immediately.

That's probably spot on..

> I believe if you did require("fzf-lua.defaults").defaults.files.fd_opts it might have also worked if you don't mind trying.

That works.. kinda. `vim.ui.select` works, but my custom keymaps don't. Probably the opts are not in a usable form yet.

I'll stick with the first suggestion as it follows what's in LazyVim already, not sure why I didn't think of checking that first.

> I would also be interested to know what is really happening, so if you don't mind pinging Folke to give us an explanation whenever he has some free time, I would welcome it 😄

I feel like I've used my annoyance allowance recently :upside_down_face: I'll take this at face value as it sounds quite logical:

>When you did require("fzf-lua").defaults you essentially load fzf-lua and access the defaults. That may be causing some conflict with lazy.nvim's setup function with fzf-lua's lazy-loading. But when you do either require("fzf-lua.config").defaults... or require("fzf-lua.defaults").defaults... you define that you access the specific module after fzf-lua has been loaded.

Thanks friend!

**Reply by dpetka2001** [reply]
> When you did require("fzf-lua").defaults you essentially load fzf-lua and access the defaults. That may be causing some conflict with lazy.nvim's setup function with fzf-lua's lazy-loading. But when you do either require("fzf-lua.config").defaults... or require("fzf-lua.defaults").defaults... you define that you access the specific module after fzf-lua has been loaded.

@dotfrag Actually, I believe that this is kinda wrong and my last comment might be more correct (although still just a guess based on a hunch) :stuck_out_tongue: 
> My final guess would be that `require("fzf-lua").defaults` accesses the `M.defaults = require("fzf-lua.defaults").defaults` which should be available after the plugin has loaded. The second approach `require("fzf-lua.config").defaults` or `require("fzf-lua.defaults").defaults` gives you direct access to the table of the specific module, which is just the original default values before any modification.

**Reply by folke** [reply]
Not really 100% sure how that could happen, but I just pushed a change that should fix this.

https://github.com/LazyVim/LazyVim/commit/33e1da585d8709bcf6c076e80a73b55ef70a9443

**Reply by dotfrag** [reply]
That fixed it indeed, thanks!
